Add install script on Windows (#1)

dmitryrec opened the pull request:
Implements #1

dmitryrec committed:
Add install script on Windows

dmitryrec committed:
Merge branch 'master' into 1-add-install-script-on-windows

dmitryrec committed:
Recreated install script on Windows

- added default path

- added 3 arguments

- added if statements

dmitryrec committed:
Add small documentation

dmitryrec committed:
Update comment syntax

dmitryrec committed:
Update if statement logic

ales-tsurko commented on install.win.bat:
```diff
@@ -0,0 +1,31 @@
+:: Usage:
+:: install.win.bat name destination_name plugin_path
```
Поменяй `plugin_path` на `plugins_dir`, а то получается, что это путь к плагину.

ales-tsurko replied:
Так ты ж не поменял)

Я говорил про коммент, это же висит к конкретной строке. Переменную, как-раз, можно было и не менять)

Еще просил показать, что это необязательный параметр, добавив его в квадратные скобки. То есть `[plugins_dir]`.

ales-tsurko commented on install.win.bat:
```diff
@@ -0,0 +1,31 @@
+
+set name=%1
+set dest_name=%2
+set PLUG_PATH=%3
```
Смотри, в моем случае в скрипте капсом были прописаны внутренние переменные, а строчными — аргументы. У тебя сейчас смешанно получается. Нужно либо все одинаково, либо чтобы какая-то логика была. Можешь их все сделать капсом, или все строчными, или которые из аргументов — строчными, а `DEFAULT_PATH` прописными (или наоборот), но если смешиваешь, это сбивает с толку.

ales-tsurko commented on install.win.bat:
```diff
@@ -0,0 +1,31 @@
+for %%x in (%*) do Set /A argsCount+=1
+
+if %argsCount%==3 (
+    rd %PLUG_PATH%\Effects\%dest_name%
```
Папку нужно удалять рекурсивно. Хз, что мы там в будущем еще будем устанавливать. Посмотри [здесь](https://stackoverflow.com/a/97896/4429188), как это делается.

ales-tsurko commented on install.win.bat:
```diff
@@ -0,0 +1,31 @@
+)
+
+if %argsCount%==2 (
+    rd %DEFAULT_PATH%\Effects\%dest_name%
+    md %DEFAULT_PATH%\Effects\%dest_name%
+
+    move target\release\examples\%name%.dll %DEFAULT_PATH%\Effects\%dest_name%\%dest_name%_x64.dll
```
Чтобы убрать дублирование целого блока (DRY), лучше в проверке по аргументам переписывать какую-то конечную переменную. Например, если ты будешь переназначать `DEFAULT_PATH`, то тебе всего лишь нужно будет добавить блок для проверки 3-х аргументов и назначать в эту переменную значение из аргумента. А сам блок вообще вынести из `if`.

То есть примерно (не знаю про batch скрипты, проверь синтаксис):

```batch
if %argsCount%==3 (
    set DEFAULT_PATH=%3
)

rd %DEFAULT_PATH%\Effects\%dest_name%
md %DEFAULT_PATH%\Effects\%dest_name%
 
move target\release\examples\%name%.dll %DEFAULT_PATH%\Effects\%dest_name%\%dest_name%_x64.dll
```

И тогда можешь `DEFAULT_PATH` переименовать в `PLUG_PATH`, а бывший `PLUG_PATH` вообще убрать.

ales-tsurko commented on install.win.bat:
```diff
@@ -0,0 +1,31 @@
+    move target\release\examples\%name%.dll %DEFAULT_PATH%\Effects\%dest_name%\%dest_name%_x64.dll
+)
+
+if %argsCount%==1 (
```
Во-первых, что если аргументов будет больше, чем нужно? Лучше проверить по "меньше двух или больше 3-х".

Во-вторых, такого рода валидацию лучше делать как можно раньше (ранний отказ), чтобы не тратилось время на все остальное... Тут, правда, не те масштабы) Но это в целом полезная привычка.

И не забудь здесь тоже изменить `plugin_path` на `plugins_dir`. Еще лучше `[plugins_dir]`, чтобы было понятно, что это необязательный аргумент.

dmitryrec replied:
это про проверку if %argsCount%==1 ?

ales-tsurko reviewed (changes requested):
Ты почти сделал это)

ales-tsurko commented on install.win.bat:
```diff
@@ -0,0 +1,31 @@
+set dest_name=%2
+set PLUG_PATH=%3
+set DEFAULT_PATH="C:\Program Files\Image-Line\FL Studio 20\Plugins\Fruity"
+set argsCount=0
```
Чуть не пропустил. Тоже самое, что и выше. Если уже используешь snake case, то `args_count`.

dmitryrec committed:
Update script structure
- rename variables to lowercase
- use snake case in all structure
- add recursive del

dmitryrec committed:
Update script structure
- make DRY

dmitryrec committed:
Update file structure
- make dry in if statements

ales-tsurko commented on install.win.bat:
```diff
@@ -0,0 +1,27 @@
+if %args_count%==3 (
+    set plugins_dir=%3
+)
+    md %plugins_dir%\Effects\%dest_name%
+    move target\release\examples\%name%.dll %plugins_dir%\Effects\%dest_name%\%dest_name%_x64.dll
```
Удаление директория, если он уже существует, решил вообще убрать, вместо того, чтобы сделать рекурсивным?)

Еще обрати внимание, у тебя тут табуляция (отступы) ненужная.

ales-tsurko commented on install.win.bat:
```diff
@@ -0,0 +1,27 @@
+if %args_count% lss 2 set flag=true
+if %args_count% gtr 3 set flag=true
+if "%flag%"=="true" (
+    echo "Usage: install.win.bat name destination_name plugin_path"
```
Здесь тоже самое про **plugins_dir**.

dmitryrec committed:
Make refactoring
- del tab
- add recursive del
- fix doc

dmitryrec committed:
Merge branch 'master' of https://github.com/tonikasoft/fpsdk into 1-add-install-script-on-windows

ales-tsurko committed:
Clean up

ales-tsurko reviewed (approved):
🎆
